File: Problem Patterns/HashTable/next-greater-element-i.py
# Problem Link: https://leetcode.com/problems/next-greater-element-i

class Solution:
    
    hashTable = {}

    def findElementsPosition(self, arr):

        for idx, num in enumerate(arr):
            self.hashTable[num] = idx
    
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        self.findElementsPosition(nums2)

        res = []

        for num in nums1:
            
            greaterFound = False
            startRange = self.hashTable[num]

            for j in range(startRange, len(nums2)):

                if nums2[j] > num:
                    res.append(nums2[j])
                    greaterFound = True
                    break
            if not greaterFound:
                res.append(-1)
        


        return res

File: Problem Patterns/HashTable/next-greater-element-i.java
// Problem Link: https://leetcode.com/problems/next-greater-element-i

class Solution {

    Map<Integer, Integer> hashTab= new HashMap<>();

    public void findElementPosition(int[] arr)
    {
        for(int i = 0; i < arr.length; i++)
        {
            this.hashTab.put(arr[i], i);
        }
    }

    public int[] nextGreaterElement(int[] nums1, int[] nums2) {

        this.findElementPosition(nums2);

        int[] res = new int[nums1.length];
        Arrays.fill(res, -1);
        int currPos = 0;
        
        for(int num: nums1)
        {
            int startingRange = this.hashTab.get(num);

            for(int i = startingRange; i < nums2.length; i++)
            {
                if(nums2[i] > num)
                {
                    res[currPos] = nums2[i];
                    break;
                }
            }
            currPos++;
        }

        return res;
    }
}
